### ISSUE TITLE
Thumbnails non-functional in Wayland using normal terminal emulators

### ISSUE THREAD

**Issue by littlediobolic** [pystardust/ytfzf#64]
Thumbnails do not display in Wayland using a normal terminal emulator due to ueberzug relying on X. This is obviously an upstream issue but something I feel is worth mentioning. For example gnome-terminal does not work.

I tested in [kitty](https://github.com/kovidgoyal/kitty)however,  which has its own image protocol and it does work, but in 99% of terminals it will likely not work by default.

**Comment by duffydack** [comment]
Nice to know kitty works, I use kitty with ranger.

**Comment by SamsiFPV** [comment]
I did a little bit of testing on this. It works on X11 based terminal emulators such as urxvt. Wayland based ones such as terminator don't work. Which has definitely to do with ueberzug being X11 based. Adding an additional option for us Wayland users would be appreciated.
PS: @littlediobolic i didn't got it to work with kitty, how did you do that? It's a wayland native app so it shouldn't work, since ytfzf only supports ueberzug and not kittys own protocol.

**Comment by napcok** [comment]
@SamsiFPV Terminator is not Wayland based, and thumbnails works very well in Terminator on X11.

**Comment by SamsiFPV** [comment]
@napcok wrong wording form my side, i meant it runs wayland natively. My bad.

**Comment by pystardust** [comment]
How about we use an external viewer? 
 I was able to make it work with sxiv (any viewer that auto updates when the file changes) . The only caveat is that it wouldn't look great since you have 2 windows, one with names and the other with thumbnails.

**Comment by arthureroberer** [comment]
maybe a pretext to re-invent ueberzug but for wayland?
could be useful for tons of things

**Comment by littlediobolic** [comment]
> PS: @littlediobolic i didn't got it to work with kitty, how did you do that? It's a wayland native app so it shouldn't work, since ytfzf only supports ueberzug and not kittys own protocol.

@SamsiFPV I did not do anything special to get it working. I just downloaded kitty, ran it, and thumbnails displayed properly. I can send logs if necessary but I am unsure what caused it to work.

**Comment by tgunnoe** [comment]
I would recommend looking at kitty https://sw.kovidgoyal.net/kitty/graphics-protocol.html

**Comment by DanKGooGLy** [comment]
What about sixel?

**Comment by Euro20179** [comment]
I tried img2sixel and lsix both didnt work but the problem is that fzf interprates input itself and doesnt let the terminal interprate the input, so sixel and anything that draws the image on the terminal itself wont work, w3m could work but the image does not stay on the screen.

**Comment by DanKGooGLy** [comment]
Try chafa

**Comment by duffydack** [comment]
would love sixel support so I can use in sway, with 'foot'.  Ranger works well with a sixel patch.  Hope this does one day

**Comment by Euro20179** [comment]
sixel just does not work in fzf, it ends up displaying the actual characters (yes I tried in a terminal that supports sixel).

**Comment by duffydack** [comment]
Shame.  Guess I'll use chafa then, and squint a bit. :)

**Comment by Euro20179** [comment]
Haha, chafa imo is actually decent but not as good as true image viewers ofc. w3m's image viewer might work with fzf but it seems to not stay on the screen for very long.

**Comment by timsofteng** [comment]
Doesn't work on `nixos` in `foot` terminal even with this flag
`ytfzf --thumb-disp-method=chafa`

**Comment by dnkl** [comment]
@timsofteng you also need `--show-thumbnails`

**Comment by timsofteng** [comment]
> @timsofteng you also need `--show-thumbnails`

Thanks!
Yes, it works with ansi format.
![image](https://user-images.githubusercontent.com/51912173/127459957-a847c281-8f81-4fd5-a3e2-10853bc106a7.png)

P.s is it limitation of fzf? Images in terminal is a big pain...
Tmux doesn't suppor sixel as wel...

**Comment by Euro20179** [comment]
> is it limitation of fzf

yes

**Comment by ankkax** [comment]
> Thumbnails do not display in Wayland using a normal terminal emulator due to ueberzug relying on X. This is obviously an upstream issue but something I feel is worth mentioning. For example gnome-terminal does not work.
> 
> I tested in [kitty](https://github.com/kovidgoyal/kitty)however, which has its own image protocol and it does work, but in 99% of terminals it will likely not work by default.

how did you get images showing with kitty, it is my main terminal and I don't get pictures.

**Comment by adworacz** [comment]
This appears to be broken currently. I just tried to use chafa on Arch, and I get an error stating that `ueberzug is not found`.

Screenshot:
![20211229_12h15m20s_grim](https://user-images.githubusercontent.com/561689/147700125-b90fd57f-5f2d-4879-a501-35cb8e6fde92.png)

**Comment by Euro20179** [comment]
> We're checking for ueberzug regardless of the thumb display method.

You are exactly correct, and I thought I fixed this but I guess not

**Comment by Euro20179** [comment]
mk well i accidently pushed the rewrite branch to master, it was pretty much done though, so it's probably a good thing because i've wanted to merge it for like 3 days now.

**Comment by adworacz** [comment]
BTW another crossenv (X11/Wayland) option may be to use `imv`, which supports IPC (with a helper utility called `imv-msg`). 

https://git.sr.ht/~exec64/imv
https://git.sr.ht/~exec64/imv/tree/master/item/doc/imv.1.txt#L315-322
https://git.sr.ht/~exec64/imv/tree/master/item/doc/imv-msg.1.txt

I might take a crack at this myself, since we'd get image display from a real image viewer.

**Comment by Euro20179** [comment]
unless I'm missing something, imv won't display ontop of the terminal, and it'll be hard to place where we want.

**Comment by Euro20179** [comment]
Alright I got something going, apply this diff to the newest commit on the master branch to test

```diff
diff --git a/ytfzf b/ytfzf
index 55660ef..28bb793 100755
--- a/ytfzf
+++ b/ytfzf
@@ -967,6 +967,12 @@ preview_start () {
 			exec 3> "$UEBERZUG_FIFO" # to keep the fifo open
 			;;
 		chafa|chafa-16|chafa-tty|catimg|catimg-256|display|w3m) : ;;
+		imv) 
+			#this will select the first video in the list to display
+			first_vid="$(jq -r '.[0].ID|select(.!=null)' < "$ytfzf_video_json_file")"
+			imv "${thumb_dir}/${fist_vid}.jpg" &
+			export imv_pid="$!"
+			;;
 		*)
 			"$thumbnail_viewer" "start" "$FZF_PREVIEW_COLUMNS" "$FZF_PREVIEW_LINES" 2>/dev/null ;;
 	esac
@@ -978,6 +984,7 @@ preview_stop () {
 			exec 3>&- # close file descriptor 3, closing ueberzug
 			;;
 		chafa|chafa-16|chafa-tty|catimg|catimg-256|display|w3m) : ;;
+		imv) kill "$imv_pid" ;;
 		*)
 			"$thumbnail_viewer" "stop" "$FZF_PREVIEW_COLUMNS" "$FZF_PREVIEW_LINES" 2>/dev/null ;;
 	esac
@@ -1062,6 +1069,10 @@ preview_display_image () {
 			dep_check "display" || die 3 "\nimagemagick is not installed\n"
 			killall display
 			display "$thumb_dir/${id}.jpg" ;;
+		imv)
+			imv-msg "$imv_pid" open "$thumb_dir/${id}.jpg"
+			imv-msg "$imv_pid" next
+			;;
 		w3m)
 			while true; do
 				printf "%b\n%s;\n" "0;1;10;130;$((FZF_PREVIEW_COLUMNS*5));$((FZF_PREVIEW_COLUMNS*3));;;;;$thumb_dir/${id}.jpg" 3 | $w3mimgdisplay_path

```

**Comment by VarLad** [comment]
@Euro20179 Any idea if something like https://github.com/pystardust/ytfzf/pull/437 can be implemented for Wezterm as well?

Wezterm has an in-built protocol for images as well
![image](https://user-images.githubusercontent.com/51269425/148432470-01aa17ab-178a-470d-9d89-2747d8d64f19.png)

**Comment by Euro20179** [comment]
Tested it out, `fzf` does not play well with `wezterm`, so it probably won't be added unless I'm missing something.

**Comment by VarLad** [comment]
@Euro20179 `kitty` option is extremely buggy

**Comment by Euro20179** [comment]
Yep, I noticed that myself, I honestly have no idea what the problem is, I haven't looked at it long but i opened an issue: #454

**Comment by Euro20179** [comment]
Ok, I think fzf gets messed up with kitty so easily because, we are using `--transfer-mode stream` (which must be done or it breaks) which sends terminal escape codes, but fzf probably also reads the terminal escape codes, because when you say press `up` that actually sends a terminal escape code.
However this still doesn't explain why the image ends up on the wrong side.

**Comment by timsofteng** [comment]
How can I try thumbnail viewing with imv?
Is it in release branch already?

**Comment by Euro20179** [comment]
Yes, but you have to update, and since the install location changed on linux, there is an extra command for this update (`sudo make uninstall-old`). (there will be a 2.1 release once I figure out what I want to do with the kitty viewier cause it's really buggy)

```sh
git clone https://github.com/ytfzf
cd ytfzf
sudo make uninstall-old
sudo make install doc
```

**Comment by timsofteng** [comment]
I've tried to use imv with last master `ytfzf` but it's pretty buggy. Sometimes it show wrong thumbnail, sometimes it even shows nothing. And quality of images are very bad (big weird pixels)

**Comment by Euro20179** [comment]
> And quality of images are very bad (big weird pixels)

This doesn't happen to me, a screehshot would be  helpful, it might be because thumbnail_quality is set to default, maybe try `ytfzf --thumbnail_quality=maxres` to get better thumbnails

> I've tried to use imv with last master ytfzf but it's pretty buggy. Sometimes it show wrong thumbnail,

This I can replicate, although inconsistently, a silmilar thing happens with kitty, (although kitty is so damn buggy), but I'm not sure how to fix it yet.

**Comment by Euro20179** [comment]
Ok, i commited (226e4c0) a possible fix for showing the wrong thumbnail. Try updating again (if you already ran `make uninstall-old` you don't have to run it again.

**Comment by timsofteng** [comment]
Works definitely better with last master.
One thing is may focus still on terminal and not jump to `imv` window?
I'm using sway. It's slightly annoying to move focus back every time on result response.

**Comment by Euro20179** [comment]
> One thing is may focus still on terminal and not jump to imv window?
> I'm using sway. It's slightly annoying to move focus back every time on result response.

This would have to be done by configuring sway.

**Comment by Euro20179** [comment]
Swayimg has been pushed to the latest commit in development branch, i'd recommend everyone try it if you're using sway.

**Comment by qoheniac** [comment]
Chafa works but the thumbnails are of extremely low quality. For Swayimg I only get a little loading symbol.

**Comment by qoheniac** [comment]
I use Sway and I see thumbnails if I run `bash ytfzf` instead of `sh` but then it shows the thumbnail in the center of the terminal and black all over the rest of the terminal. I can close the preview with Esc and navigate to the next entry and then again the thumbnail pops up as described and so on.

**Comment by Euro20179** [comment]
> I use Sway and I see thumbnails if I run bash ytfzf instead of sh

what is the output of `ls -l /bin/sh` seems it's a shell issue.

> ut then it shows the thumbnail in the center of the terminal and black all over the rest of the terminal. I can close the preview with Esc and navigate to the next entry and then again the thumbnail pops up as described and so on.

What happens when you do `swayimg -s fit -g 0,0,100,100 "path/to/some/picture/on/your/computer` & 2> /dev/null >&2`

And also what happens when you do `swaymsg -t get_tree | jq -r '..|try select(.focused==true)|[.rect.x,.rect.y,.rect.width,.rect.height,.pid]|@tsv'`

**Comment by qoheniac** [comment]
Ok, if I run it like `./ytfzf` (no thumbnail visible) and then open a second terminal and run `ps aux | grep swayimg` I get `swayimg -s fit -g 1920,767,377,263 /home/qoheniac/.cache/ytfzf/zdf-17521/thumbnails/pnWj-T9LGi8.jpg` so the image location is outside the screen coordinates. If I run that same command but replace 1920,767 with 0,0 the thumbnail shows up in the correct size in the top left corner of my screen. So the problem is that those coordinates are wrong.

**Comment by qoheniac** [comment]
I have an external monitor connected to my laptop and was working on that monitor. If I run the command in a terminal on the built-in screen the thumbnail is positioned nearly correct (about 10px too far to the left).

**Comment by qoheniac** [comment]
I guess you mean the `w_correct=$((w_half-20))` width reduction. This is fine as without it the thumbnail is 20px too wide also for me. But because of that width reduction the thumbnail center isn't correct anymore. If I replace `$((x))` with `$((x+10))` (after the `-g`) for compensation the positioning and size are both correct for me.

Regarding the second monitor (which is to the right of the built-in one), it seems as if `swaymsg -t get_tree` returns coordinates with respect to the top left corner of my left screen while swayimg interprets the coordinate with respect to the top left corner of my current screen.

**Comment by Euro20179** [comment]
> I guess you mean the w_correct=$((w_half-20)) width reduction. This is fine as without it the thumbnail is 20px too wide also for me. But because of that width reduction the thumbnail center isn't correct anymore. If I replace $((x)) with $((x+10)) (after the -g) for compensation the positioning and size are both correct for me.

Yes I do mean this, I'll check out adding 10 to x.

> Regarding the second monitor (which is to the right of the built-in one), it seems as if swaymsg -t get_tree returns coordinates with respect to the top left corner of my left screen while swayimg interprets the coordinate with respect to the top left corner of my current screen.

I honestly don't know if i can do anything about this other than look at the swayimg code, and possibly submit a pr, or look deeper at swayimg and swaymsg and see if im missing something.

**Comment by qoheniac** [comment]
If you go up the hierarchy in the sway tree until you reach an element of `.type==output` you can subtract that element's `.rect.x` and `.rect.y`from those you found for the currently focused window to get the correct position for swayimg. Unfortunately, I'm not very experienced with jq and don't have the time right now to find out how to do this properly but that way it should work.

**Comment by Euro20179** [comment]
Push has been made to development, try updating.

Another problem with this is that it ONLY works when the preview side is left. I'm gonna at least try to get it to work with preview side right. It's probably gonna get it's own function similar to `get_ueberzug_positioning`

**Comment by qoheniac** [comment]
That was quick! Why did you replace the 20 with 40? 20 looked better for me. And the way you pick the output, you always get the first screen in the tree which isn't necessarily the one with the focussed window. That's why I wrote you have to go up the hierarchy from the focussed window until you reach the output it is in. I now have no time anymore to look into this but I very much appreciate your work and will gladly test again tomorrow.

**Comment by Euro20179** [comment]
> Why did you replace the 20 with 40? 20 looked better for me

40 looks better for me lol. I'll do more testing, probably make it an option or something.

> . And the way you pick the output, you always get the first screen in the tree which isn't necessarily the one with the focussed window.

Ah I thought it was the focused one.

I"ll work on this more later

**Comment by Euro20179** [comment]
I believe it should now select the correct monitor, swaymsg gives a list of focused outputs, the output you're currently on is on top.

**Comment by qoheniac** [comment]
Yes, great! Now the positioning is correct. But that 20px shrinking still looks much better for me than the 40px. I was wondering why that might be different for your and came to think that might its an issue with font size and thus column width. So maybe we could check if it works better to do the shrinking and shifting in terms of the column width. For me it looks good if I define `TTY_COLS=$(tput cols)` and then do `w_correct=$((w_half-2*w/TTY_COLS))` (so 2 columns instead of 40px) and accordingly `swayimg -s fit -g $((x+w/TTY_COLS)),…`. Could you check is that looks good for you as well?

**Comment by Euro20179** [comment]
![image](https://user-images.githubusercontent.com/56898436/161813925-06be391a-5f0b-4c6c-806f-a95c9bd91547.png)


It's like slightly off center. but if i zoom in more it gets better.

However it's not as bad as having that -20 offset.

In addition I prefer not using `tput` since it's part of the ncurses library, however we're already using it in `interface_text`. I'll just add a little check and say you don't have tput installed.

**Comment by Euro20179** [comment]
Update:
I have basically thrown the old code out thje window as I've needed to refactor it a LOT in order to be able to accomidate for left/right/up/down previews. However it takes heavy advantage of TTY_COLS and TTY_LINES, so hopefuly it's more consistent.

It's also setting the w/h of the image without trying to preserve the aspect ratio which simplifies things a lot and it still looks fine as long as your window isn't tiny.

I'll make a push to development when I'm done, and we'll go from there.

**Comment by qoheniac** [comment]
Looking good, thank you!

**Comment by Euro20179** [comment]
> Looking good, thank you!

Excellent, I can finally consider releasing 2.2.1.
